rfcs/20230704-dynamism-101.md

```diff
@@ -0,0 +1,825 @@
+with static sizes expressed as `DenseIntElementsAttr` and dynamic sizes
+expressed as `1DTensorOf<[HLO_DimensionValue]>`.
+
+```c++
+class PadOp ... {
+  // This is how static sizes are represented in TableGen:
+  // I64ElementsAttr:$edge_padding_low
+  // I64ElementsAttr:$edge_padding_high
+  // I64ElementsAttr:$interior_padding
+  DenseIntElementsAttr getEdgePaddingLow();
+  DenseIntElementsAttr getEdgePaddingHigh();
+  DenseIntElementsAttr getInteriorPadding();
+}
+
+class DynamicPadOp ... {
+  // This is how dynamic sizes are represented in TableGen:
+  // HLO_DimensionTensor:$edge_padding_low
+  // HLO_DimensionTensor:$edge_padding_high
+  // HLO_DimensionTensor:$interior_padding
+  TypedValue<RankedTensorType> getEdgePaddingLow();
+  TypedValue<RankedTensorType> getEdgePaddingHigh();
+  TypedValue<RankedTensorType> getInteriorPadding();
+}
+```
+
+As mentioned in [F5](#f5), shape computations in StableHLO programs produce
+these 1-dimensional tensors using one of several approaches, including ops from
+Arith, Shape, StableHLO as well as other dialects. E.g. addition of dimension
+sizes can be represented via `stablehlo::AddOp`, `arith::AddIOp`, `shape::AddOp`
+and in a few other ways.
+
+Furthermore, as mentioned in [F4](#f4), using 1-dimensional tensors introduces
+considerable complexity because typical shape computations operate in terms of
+scalars. As a result, in code that constructs dynamic StableHLO ops, it is not
+unusual to encounter computations on 0-dimensional tensors whose results are
+then reshaped to 1-dimensional tensors and then concatenated together
+([example](https://github.com/tensorflow/mlir-hlo/blob/1fd13ef28d3f423363e59b1a80eb2c26e0c0979d/mhlo/transforms/chlo_legalize_to_hlo/chlo_legalize_to_hlo.cc#L1577-L1587)).
+
+**Proposal:** The RFC proposes to standardize on representing shape computations
+as StableHLO operations working on 0-dimensional tensors. More specifically,
+the proposal is to:
+
+1) Affirm that only shape computations that use StableHLO operations are
+   supported in StableHLO portable artifacts (this can be changed in future
+   RFCs, but is out of scope for this one).
+2) Use `Variadic<0DTensorOf<HLO_DimensionValue>>` instead of
+   `HLO_DimensionTensor` in the StableHLO dialect (this is consistent with
+   "variadic number of 0-dimensional tensors of integer type" from [P5](#p5)).
+3) Drop support for unranked dynamism in StableHLO (otherwise the operations
+   from P5 will no longer work for unranked operands).
+
+```tablegen
+// See P5 for the "before" and "after" of unifying static and dynamic ops.
+// This snippet only shows the "after" for this section.
+def StableHLO_PadOp ... {
+  let arguments = (ins
+    HLO_Tensor:$operand,
+    HLO_Tensor:$padding_value,
+    I64ElementsAttr:$static_edge_padding_low,
+    I64ElementsAttr:$static_edge_padding_high,
+    I64ElementsAttr:$static_interior_padding,
+    Variadic<0DTensorOf<HLO_DimensionValue>>:$dynamic_edge_padding_low,
```

Comment:
I am concerned about the usage of variadic arguments.  It can make writing [Declarative Rewrite Rule (DRR)](https://mlir.llvm.org/docs/DeclarativeRewrites/) more difficult (if not impossible).  I am not sure if this will have some impact on the [PDL dialect](https://mlir.llvm.org/docs/Dialects/PDLOps/).

If we want to go through this approach, we should think about how we can assist declarative rewrite patterns.

----

Another way I can think of to unify static/dynamic shape, is to replace both of them with `1DTensorOf<HLO_DimensionValue>` and provide a utility function that converts variadic arguments into a Tensor:

```
// Static-shape
%static_edge_padding_low = "stablehlo.constant"() {value = dense<[...]> : tensor<4xi64>} : () -> tensor<4xi64>
%static_edge_padding_high = "stablehlo.constant"() {value = dense<[...]> : tensor<4xi64>} : () -> tensor<4xi64>
%static_interior_padding = "stablehlo.constant"() {value = dense<[...]> : tensor<4xi64>} : () -> tensor<4xi64>
%pad = "stablehlo.pad"(%operand, %pad_value, %static_edge_padding_low, %static_eduge_padding_high, %static_interior_padding)

// Dynamic-shape
%edge_padding_low = "stablehlo.make_tensor"(some, value, computed, dynamically) : (i64, i64, i64, i64) -> tensor<4xi64>
%edge_padding_high = "stablehlo.make_tensor"(some, value, computed, dynamically) : (i64, i64, i64, i64) -> tensor<4xi64>
%interior_padding = "stablehlo.make_tensor"(some, value, computed, dynamically) : (i64, i64, i64, i64) -> tensor<4xi64>
%pad = "stablehlo.pad"(%operand, %pad_value, %edge_padding_low, %eduge_padding_high, %interior_padding)
```

The benefit of this approach is that the static-shape version looks similar to the prior arts (e.g. TFLite).  Though, I am not sure if this approach has been discussed or not.